base/dict.jl

```diff
@@ -266,28 +267,44 @@ end
 
 similar(d::ObjectIdDict) = ObjectIdDict()
 
+function rehash!(t::ObjectIdDict, newsz = length(t.ht))
+    t.ht = ccall(:jl_idtable_rehash, Any, (Any, Csize_t), t.ht, newsz)
+    t
+end
+
 function setindex!(t::ObjectIdDict, v::ANY, k::ANY)
+    if t.ndel >= ((3*length(t.ht))>>2)
+        rehash!(t, max(length(t.ht)>>1, 32))
+        t.ndel = 0
+    end
     t.ht = ccall(:jl_eqtable_put, Array{Any,1}, (Any, Any, Any), t.ht, k, v)
     return t
 end
 
 get(t::ObjectIdDict, key::ANY, default::ANY) =
     ccall(:jl_eqtable_get, Any, (Any, Any, Any), t.ht, key, default)
 
-pop!(t::ObjectIdDict, key::ANY, default::ANY) =
-    ccall(:jl_eqtable_pop, Any, (Any, Any, Any), t.ht, key, default)
+function pop!(t::ObjectIdDict, key::ANY, default::ANY)
+    val = ccall(:jl_eqtable_pop, Any, (Any, Any, Any), t.ht, key, secret_table_token)
```

Comment:
can we return a Nullable or flag or something, rather than having one particular sentinel value in the system that can't be put into an ObjectIdDict without causing bugs?